### ISSUE TITLE
c++ error with conversion to integer

### ISSUE THREAD

**Issue by leohaim** [serge-sans-paille/pythran#409]
Using latest anaconda python, latest boost, latest pythran and
latest clang, the following code snippet causes a compile error:

import numpy

# pythran export bug(float[])
def bug(x):

        ints=numpy.zeros(5,'int')
        ints[0]=int(numpy.floor(x))

        return ints

#print bug(2.1)

The error is as follows:

%pythran bug.py

CRITICAL Leave the vessel! Women and children first!
C++ compiler failed to compile translated code.
E: Compile error!

******** Command line was: ********
'clang' '/tmp/tmpXBA5cx.cpp' '-I/opt/anaconda/include/python2.7' '-I/opt/anaconda/lib/python2.7/site-packages/numpy/core/include/numpy' '-I/opt/anaconda/lib/python2.7/site-packages/.' '-I/opt/anaconda/lib/python2.7/site-packages/pythran' '-I/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/patch' '-DENABLE_PYTHON_MODULE' '-DUSE_GMP' '-O2' '-g' '-std=c++11' '-O2' '-shared' '-o' 'bug.so' '-L/opt/anaconda/lib/python2.7/config' '-lpthread' '-ldl' '-lutil' '-lpython2.7' '-lboost_python' '-lgmp' '-lgmpxx' '-fPIC' '-ltcmalloc_minimal'

******** Output :  ********

In file included from /tmp/tmpXBA5cx.cpp:11:
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/__builtin__/int_.hpp:13:24: error: no viable conversion from 'pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> >' to 'long'
                return t;
                       ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/__builtin__/int_.hpp:25:38: note: in instantiation of function template specialization 'pythonic::__builtin__::int_<pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> > >' requested here
        PROXY(pythonic::__builtin__, int_);
                                     ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/utils/proxy.hpp:21:35: note: expanded from macro 'PROXY'
#define PROXY(ns,f) FPROXY(ns, f, f)
                                  ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/utils/proxy.hpp:15:28: note: expanded from macro 'FPROXY'
                return ns::f(std::forward<Types>(types)...); \
                           ^
/tmp/tmpXBA5cx.cpp:66:25: note: in instantiation of function template specialization 'pythonic::__builtin__::proxy::int_::operator()<pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> > >' requested here
    std::get<0>(ints) = pythonic::__builtin__::proxy::int_{}(pythonic::numpy::proxy::floor{}(x));
                        ^
/tmp/tmpXBA5cx.cpp:77:30: note: in instantiation of function template specialization '__pythran_bug::bug::operator()<pythonic::types::ndarray<double, 1> >' requested here
  return __pythran_bug::bug()(a0);
                             ^
1 error generated.

**Comment by serge-sans-paille** [comment]
Thanks for the bug report---I can reproduce the error, it is due to the ``__builtin__.int`` function not being overloaded for ``numpy.ndarray``.

**Comment by serge-sans-paille** [comment]
(as a side note, calling ``numpy.zeros`` with a dtype given as a string is not supported, you must use ``numpy.zeros(10, numpy.int)`` instead.)

**Comment by serge-sans-paille** [comment]
There's also a typo in your code, as you're assigning a ``ndarray`` to an ``int``. So the code I used for the bugfix is

```python
import numpy
#pythran export bug(float[])

def bug(x):

    ints=numpy.zeros((5,5),numpy.int32)
    ints[0]=numpy.int32(numpy.floor(x))

    return ints
```

does that match your needs? (it turns out the ``_builtin__.int`` function does not support conversion of ndarray in Python either)

**Comment by leohaim** [comment]
Hi Serge,

yes, if this works it will fit my needs,
thanks.

After gaining some experience with pythran, I can say it produces fast 
code as expected.
For convenience of development, I have two suggestions that  would help 
me a lot if implemented:
- one should be able import self-generated pythran modules in pythran 
(see also issue Extra need of PYTHONPATH #403)
I could not get that working. If it worked this would allow splitting a 
file with many defs into smaller files and
to compile them in parallel to get better turnaround time for debugging.
Once all the split functions compile, one can compile also the whole file.
- there would be a consistency check between the
# pythran export xxxx
line and the
def xxxx
line already at the beginning when python syntax is checked.
Currently, if  an argument is missing or is an array instead of a scalar 
in the # pythran line,
this is not detected by pythran but the underlying c++ compiler, often a 
few minutes later.
If such issues were checked early before or during generation of the AST 
tree it would reduce turnaround time a lot.

Kind regards,

Leo


Am 20.03.2015 um 07:39 schrieb serge-sans-paille:
>
> There's also a typo in your code, as you're assigning a |ndarray| to 
> an |int|. So the code I used for the bugfix is
>
> import  numpy
> #pythran export bug(float[])
>
> def  bug(x):
>
>      ints=numpy.zeros((5,5),numpy.int32)
>      ints[0]=numpy.int32(numpy.floor(x))
>
>      return  ints
>
> does that match your needs? (it turns out the |_builtin__.int| 
> function does not support conversion of ndarray in Python either)
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/serge-sans-paille/pythran/issues/409#issuecomment-83938994>.
>

-- 
Ao. Prof. Dr. Leopold Haimberger
Department of Meteorology and Geophysics
Althanstrasse 14 1090 Wien
AUSTRIA
Tel. +43 1 4277 53712
Mobile: +43 664 260 5378
Fax +43 1 4277 9537
www: http://imgw.univie.ac.at/

**Comment by pbrunet** [comment]
Hi leohaim,

Thanks for these feedback.
About #403, this issue is validating before being merged in the master branch but as you say, we do not support Pythran module imported in another Pythran module. I think there are few reasons for this:
1) As it is compiled, with module `a` loading module `b` and module `c` loading module `b`, if you update `b` for something in `a`, it has some risk to break your module `b`. (Well it is not a really good argument because it may happen in Python too)
2) If we import Pythran module, it will disable some inter-procedural optimizations. (Ok, it is user choices...)
3) How can we specify if we want to import the Python module or the Pythran one? (I think this is the only real blocking issue)
4) The import of custom modules is very recent in the master branch of Pythran so we didn't have enough feedback to continue on it. But hopefully, you are here :-)

For the second point, for now, the typing part is forwarded to the C++ compiler. We plan to change this but I didn't succeed for now. I should try again soon.

Thanks
Pierrick

**Comment by leohaim** [comment]
Thanks! I will try asap.

Leo

Am 26.03.2015 um 21:19 schrieb pbrunet:
>
> Closed #409 <https://github.com/serge-sans-paille/pythran/issues/409> 
> via #410 <https://github.com/serge-sans-paille/pythran/pull/410>.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/serge-sans-paille/pythran/issues/409#event-266300403>.
>

-- 
Ao. Prof. Dr. Leopold Haimberger
Department of Meteorology and Geophysics
Althanstrasse 14 1090 Wien
AUSTRIA
Tel. +43 1 4277 53712
Mobile: +43 664 260 5378
Fax +43 1 4277 9537
www: http://imgw.univie.ac.at/
